File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head, ListNode* prev=nullptr){
        if(head==nullptr)   return prev;
        ListNode* temp = head->next;
        head->next = prev;
        return reverseList(temp, head);
    }

    ListNode* reverseKGroup(ListNode* head, int k){
        ListNode *newHead=NULL,*winPrev=NULL, *winHead=head;
        int count=1;
        while(head!=nullptr){
            ListNode* temp = head->next;
            if(count==k){
                head->next = nullptr;
                ListNode* reversed = reverseList(winHead);
                if(newHead==nullptr)    newHead=reversed;
                if(winPrev!=nullptr)    winPrev->next=reversed;
                winPrev=winHead;
                winHead=temp;
                count=1;
            }
            else   count++;
            head=temp;
        }
        winPrev->next=winHead; 
        return newHead;
    }
};

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head, prev=None):
        if not head:
            return prev
        temp = head.next
        head.next = prev
        return self.reverseList(temp, head)
    
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        newHead, winPrev, winHead = None, None, head
        count = 1
        while head:
            temp = head.next
            if count == k:
                head.next = None
                reversed_part = self.reverseList(winHead)
                if not newHead:
                    newHead = reversed_part
                if winPrev:
                    winPrev.next = reversed_part
                winPrev = winHead
                winHead = temp
                count = 1
            else:
                count += 1
            
            head = temp
        winPrev.next = winHead
        return newHead
                